Add EIP-2718(Typed Transaction Envelope) and EIP-1559 compatibility

matkt opened the pull request:
### What does this PR do?

Add EIP-2718(Typed Transaction Envelope) and EIP-1559 compatibility .

This PR allows to differentiate different types of transactions cleanly while avoiding breaking compatibility

The effort was to prepare and facilitate the implementation of new type of transaction in the future. The 1559 transactions were a good test to see if the refactor works.

Only one test has been changed (related to the last modification of the EIP1559). It's more to be sure that there is no regression

### Where should the reviewer start?
The first thing was to avoid breaking compatibility. The review can start by giving its opinion on this and how the refactor was done.

There may also be a check of how the 1559 transactions and the envelope transaction encoding have been implemented.

### Why is it needed?
With London and Berlin new EIPS were implemented. it is necessary to be compatible

mohamedelshami commented:
@matkt great timing for this PR, we will review ASAP.

AlexandrouR commented:
@matkt Great work i was wondering if at there is any client it can be tested at the moment

AlexandrouR commented:
@matkt also i think there should be an additional field `	access_list: List[Tuple[int, List[int]]] = field(default_factory=list)
`

matkt commented:
> Great work i was wondering if at there is any client it can be tested at the moment

Yes I tested with Besu to spam the calaveras network with 1559 and Legacy transactions

matkt commented:
> @matkt also i think there should be an additional field ` access_list: List[Tuple[int, List[int]]] = field(default_factory=list) `

Indeed I did not want to add the access_list (EIP-2930) to this PR because it is already a big PR with a lot of refactoring and I think it would be better to add this in another next PR IMO

AlexandrouR commented:
@matkt did you have any issues with the chain id when testing?

matkt commented:
> @matkt did you have any issues with the chain id when testing?

No. why ?  you have issues with the chaindId ?

AlexandrouR commented:
> > @matkt did you have any issues with the chain id when testing?
> 
> No. why ? you have issues with the chaindId ?

Yes i started a private besu node and i'm running the following code snippet 

```
            TransactionReceipt eip1559receipt = Transfer.sendFundsEIP1559(web3, credentials, "0xcE6D542ad2e6Bc04148f15960995C4f8B084181B", BigDecimal.valueOf(2), Convert.Unit.ETHER, Transfer.GAS_LIMIT, Transfer.GAS_LIMIT, Transfer.GAS_LIMIT).send();

```

matkt commented:
> > > @matkt did you have any issues with the chain id when testing?
> > 
> > 
> > No. why ? you have issues with the chaindId ?
> 
> Yes i started a private besu node and i'm running the following code snippet
> 
> ```
>             TransactionReceipt eip1559receipt = Transfer.sendFundsEIP1559(web3, credentials, "0xcE6D542ad2e6Bc04148f15960995C4f8B084181B", BigDecimal.valueOf(2), Convert.Unit.ETHER, Transfer.GAS_LIMIT, Transfer.GAS_LIMIT, Transfer.GAS_LIMIT).send();
> ```

have you enabled the london fork in your testnet ?

matkt commented:
> @matkt did you have any issues with the chain id when testing?

I think I got it in my test with besu I called directly the RawManager class
```
sign(
        RawTransaction.createEtherTransaction(
                    eip1559Transaction.chainId(),
                    eip1559Transaction.nonce(),
                    DEFAULT_GAS_LIMIT,
                    eip1559Transaction.recipientAddress(),
                    Convert.toWei(eip1559Transaction.value(), Convert.Unit.ETHER).toBigInteger(),
                    eip1559Transaction.gasPremium(),
                    eip1559Transaction.feeCap()),
        credentials);
```
It seems that the RawManager is initialized without chaindId in the `sendFundsEIP1559` method. I will fix that asap

matkt commented:
@AlexandrouR I updated the code to get the chaindId from the web3J instance. Feel free to tell if it's fine for you now

ligi commented on crypto/src/test/java/org/web3j/crypto/TransactionEncoderTest.java:
```diff
@@ -93,7 +93,7 @@ public void testEip1559Transaction() {
                 TransactionEncoder.signMessage(
                         createEip1559RawTransaction(), (byte) 1, credentials),
```
I think passing 1 as chainId here is a bit weird and the chainID coming from createEip1559RawTransaction is 1559 - this feels really odd

matkt replied:
Yes thanks I updated the test to have a consistent usecase

ligi commented on crypto/src/main/java/org/web3j/crypto/transaction/type/Transaction1559.java:
```diff
@@ -0,0 +1,139 @@
+
+/**
+ * Transaction class used for signing 1559 transactions locally.<br>
+ * For the specification, refer to p4 of the <a href="http://gavwood.com/paper.pdf">yellow
```
I think for 1559 transactions this should link to EIP1559 - nothing about that in the yellow paper

matkt replied:
Yes good catch I updated that

ligi commented on crypto/src/test/java/org/web3j/crypto/TransactionEncoderTest.java:
```diff
@@ -93,7 +93,7 @@ public void testEip1559Transaction() {
                         createEip1559RawTransaction(), (byte) 1, credentials),
                 (Numeric.hexStringToByteArray(
-                        "0xf866808082753094627306090abab3a6e1400e9345bc60c78a8bef577b8082162e8310c8e026a05b362f1b948135039788c28613e8009c4d38346ece108ebc2f856902d2133243a07f42ab9c3d50e2f6cce3360cba206c7a4065bac0cfca015611af8be7189730fe")));
+                        "02f8698206178082162e8310c8e082753094627306090abab3a6e1400e9345bc60c78a8bef577b80c00aa083cac99935549adc82c604e28635c8dc908956d822b51e7cf843cc91b23bc4e3a04cef4ed7a7d58581194a840aa4f311b4d74f6a42b7bfd73b08336dbb2e424dd2")));
```
I am a bit puzzled about this test-case as v is 10 here and I would expect it to be 1

matkt replied:
V is equal to 27 and it's normal because in the yellow paper you can see that the range of V is [27,30]. You can check the value of V here https://github.com/matkt/web3j/blob/f01f3dec17e45b6e0f44f3fe115f08356b5159be/crypto/src/main/java/org/web3j/crypto/TransactionEncoder.java#L45 with `signatureData.getV()[0]`

Then for 1559 transactions this is the recId that we put and not V.  If we have 27 for V so I expect a recId of 0 but maybe I missed something. 

But there is an error because I did not take into account the chainId to find the RecId. We need to reverse that https://github.com/web3j/web3j/blob/49fe2c4e2d9d325ec465879736d6c384f41a4115/crypto/src/main/java/org/web3j/crypto/TransactionEncoder.java#L60 to retrieve the good value from the signature. So I will updated that and I will retry with Besu to be sure it's fine

ligi replied:
yea I mean 1 for recId - so in the place of V - but it was 10 when I decoded the RLP

ligi commented:
Thanks for the changes - after 9325400 the test-case also passes against the KEthereum implementation

matkt commented:
> Thanks for the changes - after [9325400](https://github.com/web3j/web3j/commit/93254009a5e531258873aab49b2189219beaa32a) the test-case also passes against the KEthereum implementation

cool thanks for your review. I will test to send transactions on calaveras in order to validate that everything is still good

ligi commented:
I am actually a bit surprised It worked before ;-)
But should be good now I tested my implementation on calaveras

matkt commented:
> I am actually a bit surprised It worked before ;-)
> But should be good now I tested my implementation on calaveras

yes it's quite strange knowing that the spammed tests were done with this version 🤔

AlexandrouR commented:
@matkt  All good tested it locally on besu master with the following snippet 
```
    TransactionReceipt transactionReceipt = Transfer.sendFundsEIP1559(
                    web3, credentials, "0xcE6D542ad2e6Bc04148f15960995C4f8B084181B",
                    BigDecimal.valueOf(2),
                    Convert.Unit.ETHER,
                    DefaultGasProvider.GAS_LIMIT,
                    DefaultGasProvider.GAS_LIMIT,
                    DefaultGasProvider.GAS_PRICE)
                    .send();
```
and it went through.
What worries me a bit at this point is that if the gas values are no configured properly I noticed that while the client's console will complain eg ` Invalid transaction: max priority fee per gas cannot be greater than max fee per gas ` web3j continues to poll for the receipt.

matkt commented:
> All good tested it locally on besu master with the following snippet

Thanks for your feedback. Do you have the same behavior with an invalid legacy transaction ?

AlexandrouR commented:
> > All good tested it locally on besu master with the following snippet
> 
> Thanks for your feedback. Do you have the same behavior with an invalid legacy transaction ?

With an invalid legacy transaction, I get this ` Error processing transaction request: Intrinsic gas exceeds gas limit ` if the gas values are wrong. The way I see it it could be two things i guess: The client sends the error message but web3j can't read it and it keeps polling, or the client does not send the error message back.

matkt commented:
> > > All good tested it locally on besu master with the following snippet
> > 
> > 
> > Thanks for your feedback. Do you have the same behavior with an invalid legacy transaction ?
> 
> With an invalid legacy transaction, I get this `Error processing transaction request: Intrinsic gas exceeds gas limit` if the gas values are wrong. The way I see it it could be two things i guess: The client sends the error message but web3j can't read it and it keeps polling, or the client does not send the error message back.

I tried and yes I think the issue come from Besu. Because when I send a transaction with "max priority fee per gas greater than max fee per gas" I don't have an error . The good behavior will be to return an error like other invalid transaction testcases.
I will fix that on Besu 👍

AlexandrouR commented:
Referencing the Besu [PR](https://github.com/hyperledger/besu/pull/2424) here so we can track it and test once it is merged.

AlexandrouR commented:
> Referencing the Besu [PR](https://github.com/hyperledger/besu/pull/2424) here so we can track it and test once it is merged.

PR was merge i will test again later today with Besu master

matkt commented:
> > Referencing the Besu [PR](https://github.com/hyperledger/besu/pull/2424) here so we can track it and test once it is merged.
> 
> PR was merge i will test again later today with Besu master

did the PR on Besu fix your issue ?

AlexandrouR commented:
@matkt  Yes all good I have tested a few scenarios and web3j received the errors as it should.

AlexandrouR commented:
The issue now is with the geth client as i get the following error `Error processing transaction request: invalid sender`

matkt commented:
> invalid sender

what is the network where you are trying to send the transaction  ?

AlexandrouR commented:
> > invalid sender
> 
> what is the network where you are trying to send the transaction ? still calaveras ?

no private network with the London fork enabled. the legacy transactions are going through normally.

matkt commented:
> > > invalid sender
> > 
> > 
> > what is the network where you are trying to send the transaction ? still calaveras ?
> 
> no private network with the London fork enabled. the legacy transactions are going through normally.

It think it's an  invalid chainId issue. Are you using the chainId of your private network ?

AlexandrouR commented:
> > > > invalid sender
> > > 
> > > 
> > > what is the network where you are trying to send the transaction ? still calaveras ?
> > 
> > no private network with the London fork enabled. the legacy transactions are going through normally.
> 
> It think it's an  invalid chainId issue. Are you using the chainId of your private network ?

Yes but the chain id is obtained by calling  client for it.

vae520283995 commented:
@matkt 
```
public static RawTransaction createEtherTransaction(
            BigInteger chaindId,
            BigInteger nonce,
            BigInteger gasLimit,
            String to,
            BigInteger value,
            BigInteger maxPriorityFeePerGas,
            BigInteger maxFeePerGas) {
        return new RawTransaction(
                Transaction1559.createEtherTransaction(
                        chaindId, nonce, gasLimit, to, value, maxPriorityFeePerGas, maxFeePerGas));
    }
```
have a typo, should be chainId

vae520283995 commented:
> The issue now is with the geth client as i get the following error `Error processing transaction request: invalid sender`

I have the same problem, I am using ropsten network, chainId is 3

matkt commented:
> result.add(RlpString.create(Sign.getRecId(signatureData, chainId)));

Looking at this implementation https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1559.md
For EIP1559 it's no longer V but signatureYParity. It's working with Besu

For Geth I tried to send a 1559 transaction via Infura to the Ropsten network. The instance used by Infura it's `Geth/v1.10.4-omnibus-3dd33368/linux-amd64/go1.16.5` and it's fine the transaction are mined on a block.

Could you show me how are you sending the transaction via web3j ? To help me reproduce the issue

vae520283995 commented:
> Could you show me how are you sending the transaction via web3j ? To help me reproduce the issue

```
RawTransaction eip1559Transaction = RawTransaction.createEtherTransaction(
                BigInteger.valueOf(ChainIdLong.ROPSTEN),
                BigInteger.valueOf(1),
                BigInteger.valueOf(21000),
                "0xa8eC6970FEc8842Dd229EEf170f624c3D77DAb45",
                Convert.toWei(BigDecimal.valueOf(0.1), Convert.Unit.ETHER).toBigInteger(),
                Convert.toWei(BigDecimal.valueOf(1), Convert.Unit.GWEI).toBigInteger(),
                Convert.toWei(BigDecimal.valueOf(2), Convert.Unit.GWEI).toBigInteger()
        );
        ECKeyPair ecKeyPair = ECKeyPair.create(new BigInteger(privateKey, 16));
        Credentials credentials = Credentials.create(ecKeyPair);
//        byte[] signMessage = TransactionEncoder.signMessage(eip1559Transaction, credentials);
        byte[] signMessage = TransactionEncoder.signMessage(eip1559Transaction, ChainIdLong.ROPSTEN, credentials);
        System.out.println(Numeric.toHexString(signMessage));
```

use
```
byte[] signMessage = TransactionEncoder.signMessage(eip1559Transaction, credentials);
```
ok

use
```
 byte[] signMessage = TransactionEncoder.signMessage(eip1559Transaction, ChainIdLong.ROPSTEN, credentials);
```
throw
```
Exception in thread "main" java.lang.RuntimeException: Unsupported format exception
	at org.web3j.crypto.Sign.getRecId(Sign.java:286)
	at org.web3j.crypto.transaction.type.Transaction1559.asRlpValues(Transaction1559.java:89)
	at org.web3j.crypto.TransactionEncoder.asRlpValues(TransactionEncoder.java:113)
	at org.web3j.crypto.TransactionEncoder.encode(TransactionEncoder.java:93)
	at org.web3j.crypto.TransactionEncoder.encode(TransactionEncoder.java:84)
	at org.web3j.crypto.TransactionEncoder.signMessage(TransactionEncoder.java:47)
	at org.web3j.Eip1559.main(Eip1559.java:47)

FAILURE: Build failed with an exception.
```

I know what the problem is, you should add a judgment in 
public static byte[] signMessage(RawTransaction rawTransaction, long chainId, Credentials credentials)
```
public static byte[] signMessage(
            RawTransaction rawTransaction, long chainId, Credentials credentials) {
        if (!rawTransaction.getType().equals(TransactionType.LEGACY)) {
            return signMessage(rawTransaction, credentials);
        }
        byte[] encodedTransaction = encode(rawTransaction, chainId);
        Sign.SignatureData signatureData =
                Sign.signMessage(encodedTransaction, credentials.getEcKeyPair());

        Sign.SignatureData eip155SignatureData = createEip155SignatureData(signatureData, chainId);
        return encode(rawTransaction, eip155SignatureData);
    }
```

matkt commented:
> if (!rawTransaction.getType().equals(TransactionType.LEGACY)) {
>             return signMessage(rawTransaction, credentials);
>         }

Thanks for your feedback I updated the code and I changed the chainId type. Let me know if you any other issues

AlexandrouR commented:
@matkt the issue is fixed. Thank you. i believe this PR is ready to be merged.

AlexandrouR commented:
@matkt  could you apply spotless please so we can merged ?
